libraries/program-error/tests/mod.rs

```diff
@@ -0,0 +1,135 @@
+pub mod bench;
+pub mod decode;
+pub mod into;
+pub mod print;
+pub mod spl;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use solana_program::{
+        decode_error::DecodeError,
+        program_error::{PrintProgramError, ProgramError},
+    };
+    use std::sync::{Arc, RwLock};
+
+    // Used to capture output for `PrintProgramError` for testing
+    lazy_static::lazy_static! {
+        static ref EXPECTED_DATA: Arc<RwLock<Vec<u8>>> = Arc::new(RwLock::new(Vec::new()));
+    }
+    fn set_expected_data(expected_data: Vec<u8>) {
+        *EXPECTED_DATA.write().unwrap() = expected_data;
+    }
+    pub struct SyscallStubs {}
+    impl solana_sdk::program_stubs::SyscallStubs for SyscallStubs {
+        fn sol_log(&self, message: &str) {
+            assert_eq!(
+                message,
+                String::from_utf8_lossy(&*EXPECTED_DATA.read().unwrap())
+            );
+        }
+    }
+
+    // `#[derive(IntoProgramError)]`
+    #[test]
+    fn test_derive_into_program_error() {
+        // `Into<ProgramError>`
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::MintHasNoMintAuthority),
+            Into::<ProgramError>::into(into::ExampleError::MintHasNoMintAuthority),
+        );
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::IncorrectMintAuthority),
+            Into::<ProgramError>::into(into::ExampleError::IncorrectMintAuthority),
+        );
+    }
+
+    // `#[derive(DecodeError)]`
+    #[test]
+    fn test_derive_decode_error() {
+        // `Into<ProgramError>`
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::MintHasNoMintAuthority),
+            Into::<ProgramError>::into(decode::ExampleError::MintHasNoMintAuthority),
+        );
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::IncorrectMintAuthority),
+            Into::<ProgramError>::into(decode::ExampleError::IncorrectMintAuthority),
+        );
+        // `DecodeError<T>`
+        assert_eq!(
+            <bench::ExampleError as DecodeError<bench::ExampleError>>::type_of(),
+            <bench::ExampleError as DecodeError<decode::ExampleError>>::type_of(),
+        );
+    }
+    // `#[derive(PrintProgramError)]`
+    #[test]
+    fn test_derive_print_program_error() {
```

Comment:
To avoid segfaulting or other potential false negatives, this test and the following should be executed `#[serial]`ly